http-tests/src/test/scala/org/apache/pekko/http/scaladsl/server/directives/MarshallingDirectivesSpec.scala

```diff
@@ -275,4 +275,23 @@ class MarshallingDirectivesSpec extends RoutingSpec with Inside {
       }
     }
   }
+
+  "The marshalling infrastructure for text" should {
+    val foo = "Hällö"
+    "render text with UTF-8 encoding if no `Accept-Charset` request header is present" in {
+      Get() ~> complete(foo) ~> check {
+        responseEntity shouldEqual HttpEntity(ContentType(`text/plain`, `UTF-8`), foo)
+      }
+    }
+    "render text with requested encoding if an `Accept-Charset` request header requests a non-UTF-8 encoding" in {
+      Get() ~> `Accept-Charset`(`ISO-8859-1`) ~> complete(foo) ~> check {
+        responseEntity shouldEqual HttpEntity(ContentType(`text/plain`, `ISO-8859-1`), foo)
+      }
+    }
+    "render text with UTF-8 encoding if an `Accept-Charset` request header requests an unknown encoding" in {
+      Get() ~> `Accept-Charset`(HttpCharset("unknown")(Nil)) ~> complete(foo) ~> check {
+        responseEntity shouldEqual HttpEntity(ContentType(`text/plain`, `UTF-8`), foo)
+      }
+    }
```

Comment:
I realize this is a comment on the current behavior, and not on your change, but this seems strange to me: it looks like the default string marshaller will return the string passed to it as `text/plain` claiming it is in whatever encoding that was requested by the user, without doing anything to make sure the string is actually in that encoding.

This is only correct if you assume the implementation correctly took into account the charset, which seems extremely unlikely. I wonder if it wouldn't make more sense to leave out the optional `charset` parameter to the return content type in this case. If people have specific requirements (which seems somewhat unlikely) they can specify the charset explicitly.

This would be a behavior change, but it seems like a reasonable one if we note it in the release notes.

Reply:
before the change to the main source code in this PR, this test actually fails - the response is an error response

Reply:
Hi, 

Thanks for having a look at this bug / feature request. I recently sent in a duplicate issue (https://github.com/apache/pekko-http/issues/583) unaware of this one. I worked on a similar PR the past two days, and have now found and looked through your changes - that would seem to fix it nicely! Great. 

For your consideration, I'd like to suggest adding in an extra unit-test in "MarshallingSpac.scala" with something along the lines of:

```
val invalidAcceptCharsetHeader = `Accept-Charset`(HttpCharsetRange(HttpCharset.custom("invalid")))
val request = HttpRequest().withHeaders(Seq(invalidAcceptCharsetHeader))
marshalToResponse("Hello", request).entity.contentType.charsetOption should be(Some(HttpCharsets.`UTF-8`))
```

Some refactoring thoughts; Perhaps the try-catch in "PredefinedToEntityMarshallers.scala" could be more subtly placed in "HttpCharset.scala" as:

```
/** Returns this HttpCharset instance if the provided nio charset is available or a desired default otherwise */
  def withNioCharsetOrElse(default: HttpCharset): HttpCharset = {
    if (_nioCharset.isSuccess) {
      // Return this instance, as the provided nioCharset did not result in a java.nio.charset.UnsupportedCharsetException
      this
    } else {
      default
    }
  }
```

and then have the "PredefinedToEntityMarshallers.scala" do:

```
Marshaller.withOpenCharset(mediaType) { (s, cs) => {
  HttpEntity(mediaType.withCharset(cs.withNioCharsetOrElse(HttpCharsets.`UTF-8`)), s)
} }
```

Looking forward to seeing your changes merged.

Thanks again and best regards,
Michael

Reply:
Thanks @michaelf-crwd - I have added a new HttpCharsets function similar to what you suggested (c8514e78919d4bc41cc459aee9cb88a0a94043a6).
I will look later at adding extra tests like the one you suggested.

Reply:
@raboof would you be able to take a look at this again? I don't think this change breaks anything and makes the pekko-http support for `Accept-Charset` a bit more tolerant.

Reply:
@mdedetrich would you have time to look at this? It's basically the last item to decide on before a 1.1.0 release.

Reply:
Ill check this tomorrow, there is context I need to grock

Reply:
> without doing anything to make sure the string is actually in that charset

@raboof, WDYM with that? Java strings are basically UTF-16 encoded, so they are known to be unicode. What else could go wrong?

Reply:
what I meant is, when the route does `complete(some_unicode_string)`, and the marshalling logic results in a `Content-Type: text/plain; charset=ISO-8859-1` header, that response seems wrong: the header promises the body will be in ISO-8859-1, but the body is actually in unicode. I think it would make more sense to make sure we produce a `Content-Type: text/plain; charset=UTF-8` header unless the developer explicitly marked the response as being in a different charset.

Reply:
or will the marshalling logic actually convert the unicode to the charset from the content type? in that case of course it's all good.

Reply:
Indeed it does, sorry about the noise :facepalm:

Reply:
:+1: